src/CurlOps.cc

```diff
@@ -149,24 +152,51 @@ CurlOperation::StartBroker(std::string &err)
     return true;
 }
 
+bool
+CurlOperation::HeaderTimeoutExpired() {
+    if (m_received_header) return false;
+
+    struct timespec now;
+    if (clock_gettime(CLOCK_MONOTONIC, &now) == -1) {
+        return false;
+    }
+
+    auto res = (now.tv_sec > m_header_expiry.tv_sec || (now.tv_sec == m_header_expiry.tv_sec && now.tv_nsec > m_header_expiry.tv_nsec));
+    if (res) {
+        m_error = OpError::ErrHeaderTimeout;
+    }
+    return res;
+}
+
 void
 CurlOperation::Setup(CURL *curl)
 {
     if (curl == nullptr) {
         throw std::runtime_error("Unable to setup curl operation with no handle");
     }
-    m_curl.reset(curl);
-    if (m_timeout == 0) {
-        curl_easy_setopt(m_curl.get(), CURLOPT_TIMEOUT, 30);
-    } else {
-        curl_easy_setopt(m_curl.get(), CURLOPT_TIMEOUT, m_timeout);
+    struct timespec now;
+    if (clock_gettime(CLOCK_MONOTONIC, &now) == -1) {
+        throw std::runtime_error("Unable to get current time");
     }
-    curl_easy_setopt(m_curl.get(), CURLOPT_TIMEOUT, 7);
+    if (m_header_timeout.tv_sec == 0 && m_header_timeout.tv_nsec == 0) {
+        m_header_timeout = {30, 0};
+    }
+    m_header_expiry.tv_sec = now.tv_sec + m_header_timeout.tv_sec;
+    m_header_expiry.tv_nsec = now.tv_nsec + m_header_timeout.tv_nsec;
+    while (m_header_expiry.tv_nsec > 1'000'000'000) {
+        m_header_expiry.tv_nsec -= 1'000'000'000;
+        m_header_expiry.tv_sec ++;
+    }
```

Comment:
Honestly, it's a bit of defensive programming.  I was worried about `m_header_timeout` being normalized incorrectly, causing the sum `now.tv_nsec + m_header_timeout.tv_nsec` to be greater than `2e9`.

That can't happen in the current code because the other places that generate `m_header_timeout` _should_ be correct.  Just wanted to be more robust against a potential future bug.